NetKet/Machine/ffnn.hpp

```diff
@@ -0,0 +1,248 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Eigen/Dense>
+#include <iostream>
+#include <vector>
+#include "Lookup/lookup.hpp"
+#include "Utils/all_utils.hpp"
+#include "layer.hpp"
+
+#ifndef NETKET_FFNN_HPP
+#define NETKET_FFNN_HPP
+
+namespace netket {
+
+template <typename T>
+class FFNN : public AbstractMachine<T> {
+  using VectorType = typename AbstractMachine<T>::VectorType;
+  using MatrixType = typename AbstractMachine<T>::MatrixType;
+
+  std::vector<std::unique_ptr<AbstractLayer<T>>>
+      layers_;  // Pointers to hidden layers
+
+  std::vector<int> layersizes_;
+  int depth_;
+  int nlayer_;
+  int npar_;
+  int nv_;
+
+  int mynode_;
+
+  const Hilbert &hilbert_;
+
+ public:
+  using StateType = typename AbstractMachine<T>::StateType;
+  using LookupType = typename AbstractMachine<T>::LookupType;
+
+  // constructor
+  explicit FFNN(const Hilbert &hilbert, const json &pars)
+      : nv_(hilbert.Size()), hilbert_(hilbert) {
+    from_json(pars);
```

Comment:
Calling virtual functions from a constructor should be avoided (see https://google.github.io/styleguide/cppguide.html#Doing_Work_in_Constructors and https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-ctor-virtual for the reasons).

You could use a factory function instead (i.e., a static function initializing the Machine and then returning it).

Reply:
I agree, but personally I find factory logic a bit too involved... I wonder if there is some easier alternative

Reply:
The case where problems with virtual function calls from the constructor could occur is when a subclass of `FFNN` overrides `from_json`. (This constructor would still call `FFNN::from_json`,  not the version in the subclass. That might even be the intended behaviour but can still be confusing.)

If `FFNN` is never subclassed, this is not a problem, so I think marking the whole class as `final` would be an alternative to changing the code here, i.e.,
```cpp
class FFNN final : public AbstractMachine<T> {
```
(cf. the second link in my comment above).

Generally, I think this discussion is a symptom of the fact that in the current design we are strongly coupling the initialization of our classes to the JSON representation which tends to lead to difficulties in terms of code design. However, this is not specific to this PR and changing things in this regard should be done consistently for all of netket, so I'd say this is a topic for another time.